# 02.4 선형 연립방정식과 역행렬
- https://datascienceschool.net/view-notebook/927d7f7972dd434ead9d294169ae7f34/
- 선형 예측 모형은 입력 데이터 벡터와 가중치 벡터의 내적으로 계산된 예측값이 실제 출력 데이터와 유사한 값을 출력하도록 하는 모영을 말한다.
- 과거 입력 데이터들과 그 입력 데이터들의 실제 출력 데이터와의 관계를 통해 가중치 벡터를 구해야 하며 여기서 선형연립방정식과 역행렬의 개념이 필요하다.
- 이러한 과정을 거쳐서 가중치 벡터를 구해내는 과정을 선형 예측 모형을 알아내는 과정, 즉 모델링 과정이라고 말할 수 있다.

## 선형 연립방정식; system of linear equations
- 복합의 미지수를 포함하는 복합의 선형방정식

\begin{align}
\begin{matrix}
a_{11} x_1 & + \;& a_{12} x_2   &\; + \cdots + \;& a_{1M} x_M &\; = \;& b_1 \\
a_{21} x_1 & + \;& a_{22} x_2   &\; + \cdots + \;& a_{2M} x_M &\; = \;& b_2 \\
\vdots\;\;\; &   & \vdots\;\;\; &                & \vdots\;\;\; &     & \;\vdots \\
a_{N1} x_1 & + \;& a_{N2} x_2   &\; + \cdots + \;& a_{NM} x_M &\; = \;& b_N \\
\end{matrix}
\tag{2.4.2}
\end{align}

\begin{align}
\begin{bmatrix}
a_{11} & a_{12} & \cdots & a_{1M} \\
a_{21} & a_{22} & \cdots & a_{2M} \\
\vdots & \vdots & \ddots & \vdots \\
a_{N1} & a_{N2} & \cdots & a_{NM} \\
\end{bmatrix}
\begin{bmatrix}
x_1 \\ x_2 \\ \vdots \\ x_M
\end{bmatrix}
=
\begin{bmatrix}
b_1 \\ b_2 \\ \vdots \\ b_N
\end{bmatrix}
\tag{2.4.3}
\end{align}

\begin{align}
A = 
\begin{bmatrix}
a_{11} & a_{12} & \cdots & a_{1M} \\
a_{21} & a_{22} & \cdots & a_{2M} \\
\vdots & \vdots & \ddots & \vdots \\
a_{N1} & a_{N2} & \cdots & a_{NM} \\
\end{bmatrix}
, \;\;
x = 
\begin{bmatrix}
x_1 \\ x_2 \\ \vdots \\ x_M
\end{bmatrix}
, \;\;
b=
\begin{bmatrix}
b_1 \\ b_2 \\ \vdots \\ b_N
\end{bmatrix}
\tag{2.4.4}
\end{align}

\begin{align}
Ax = b  
\tag{2.4.5}
\end{align}

- A: 계수행렬 (coefficient matrix) --> 입력 데이터 매트릭스
- x: 미지수벡터 (unknown vector) --> 가중치 벡터
- b: 상수벡터 (constatnt vector) --> 출력 데이터 벡터

##### $Ax = b$ 에서 $x$를 구하는 법
- $A, x, b$ 가 스칼라 실수라면, $x = \dfrac{b}{A}$

- $A, x, b$ 가 벡터/행렬이라면, $x = A^{-1}b$
    - 여기서 역행렬의 개념이 필요하게 됨
    - $x$, 즉 가중치 벡터를 구해내면 모델링을 완료한 것

----

## 역행렬; inverse matrix

- 행렬 A의 역행렬은 본래의 행렬 A와 곱했을 때 그 결과를 항등행렬로 만드는 행렬을 말한다.
    - 행렬과 행렬의 곱이 항등행렬(Identity Matrix)이 되려면 그 두개의 행렬은 "정방행렬(square matrix)"이어야 한다.

\begin{align}
A^{-1} A = A A^{-1} = I  
\tag{2.4.9}
\end{align}

 - 역행렬은 존재할 수도, 존재하지 않을 수 있다.
    - 역행렬이 존재하는 행렬: 가역행렬(invertible matrix), 정칙행렬(regular matrix), 비특이행렬(signular matrix)
    - 역행렬이 존재하지 않는 행렬: 비가역행렬(non-invertible matrix), 특이행렬(singular matrix), 퇴화행렬(degenerate matrix)

### 역행렬의 성질

1. $(A^{T})^{-1} = (A^{-1})^{T}$


2. $(AB)^{-1} = B^{-1} A^{-1} $
    - $(ABC)^{-1} = C^{-1} B^{-1} A^{-1} $
    - cf. $(AB)^T = B^T A^T$

### 역행렬의 계산

\begin{align}
A^{-1} = \dfrac{1}{\det (A)} C^T = \dfrac{1}{\det (A)} 
\begin{bmatrix}
C_{1,1} & \cdots & C_{N,1}  \\
\vdots  & \ddots & \vdots   \\
C_{1,N} & \cdots & C_{N,N}  \\
\end{bmatrix}
\tag{2.4.14}
\end{align}

- $C_{i,j}$: 행렬 $A$의 $i,j$번째 원소에 대해 정의한 Cofactor를 의미함
- $C$: cofactor들로 이루어진 행렬을 의미함
    - 여인수행렬(matrix of cofactors, cofactor matrix)
- $C^T$: 여인수행렬의 전치행렬을 의미함 
    - 수반행렬, 어드조인트행렬(adjoint matrix, adjugate matrix)
    - $adj(A)$

- 위 식에서 $det(A)$가 0이면 $A^{-1}$은 존재하지 않느다. (분모가 0인 수는 존재하지 않으므로)

### 역행렬에 대한 몇가지 정리
- '이런게 있었었지' 라고 생각해낼 수 있으면 될 뿐, 외울 필요는 없다.

1. 셔먼-모리슨(Sherman–Morrison) 공식
    \begin{align}
    (A+uv^T)^{-1} = A^{-1} - {A^{-1}uv^T A^{-1} \over 1 + v^T A^{-1}u} 
    \tag{2.4.21}
    \end{align}
    - $A$가 조금 변하면 $A^{-1}$도 조금씩 변하는데, 그 변하는 비중에 대해 알 수 있음

2. 우드베로(woodbury) 공식
    \begin{align}
    \left(A+UCV \right)^{-1} = A^{-1} - A^{-1}U \left(C^{-1}+VA^{-1}U \right)^{-1} VA^{-1} 
    \tag{2.4.22}
    \end{align}
    - 셔먼-모리슨 공식의 약간 develop 버전

3. 분할행렬(partitional matrix)의 역행렬
    \begin{align}
    \begin{bmatrix}
    A_{11} & A_{12} \\
    A_{21} & A_{22} 
    \end{bmatrix}^{-1}
    =
    \begin{bmatrix}
    A_{11}^{-1}(I + A_{12}FA_{11}^{-1}) & -A_{11}^{-1}A_{12}F \\
    -FA_{21}A_{11}^{-1} & F
    \end{bmatrix}
    \tag{2.4.23}
    \end{align}
    

\begin{align}
F = (A_{22} - A_{21}A_{11}^{-1}A_{12})^{-1}
\tag{2.4.24}
\end{align}



$$or$$



\begin{align}
F = (A_{11} - A_{12}A_{22}^{-1}A_{21})^{-1}
\tag{2.4.25}
\end{align}

### numpy를 통한 역행렬 계산하기

In [1]:
import numpy as np

In [2]:
A = np.array([[1,1,0],[0,1,1],[1,1,1]]) #임의의 행렬 A 정의
A 

array([[1, 1, 0],
       [0, 1, 1],
       [1, 1, 1]])

In [3]:
Ainv = np.linalg.inv(A) # numpy의 linalg 서브 패키지를 이용해서 invert matrix 구하기
Ainv

array([[ 0., -1.,  1.],
       [ 1.,  1., -1.],
       [-1.,  0.,  1.]])

In [4]:
Ainv@A # 어느 행렬과 그 역행렬의 곱은 항상 identity matrix 이다.

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [5]:
A@Ainv

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

----

## 선형 연립방정식과 선형 예측모형

- 선형 예측모형의 가중치벡터를 구하는 문제는 선형 연립방정식 문제를 푸는 것과 같다.
    - 선형 연립방정식: $ Ax = b$
    - 선형예측모형: $ Xw = \hat{y}$

\begin{align}
\begin{matrix}
x_{11} w_1 & + \;& x_{12} w_2   &\; + \cdots + \;& x_{1N} w_N &\; = \;& y_1 \\
x_{21} w_1 & + \;& x_{22} w_2   &\; + \cdots + \;& x_{2N} w_N &\; = \;& y_2 \\
\vdots\;\;\; &   & \vdots\;\;\; &                & \vdots\;\;\; &     & \;\vdots \\
x_{N1} w_1 & + \;& x_{N2} w_2   &\; + \cdots + \;& x_{NN} w_N &\; = \;& y_N \\
\end{matrix}
\tag{2.4.30}
\end{align}

### numpy를 이용해서 선형 연립방정식 풀기

$$Xw = \hat{y}$$


$$w = X^{-1}\hat{y}$$

- 선형연립방정식을 풀기 위해서 inv()를 구한 후 벡터 y에 직접 곱해도 되겠지만, 보통은 lstsq() 명령을 사용한다.
- inv() 명령보다 lstsq() 명령이 수치오자도 적고, 코드도 간단하다.

- lstsq() 명령은 행렬 A와 b를 인수로 받고, "최소차승문제(least sqaure problem)"을 푸는 함수다,
    - x: 답
    - resid: 잔차제곱합(residual sum of sqaures)
    - rank: 랭크 - 행렬 $X$가 얼마나 독립적인 데이터, 변수였는지에 대한 여부 > 높을 수록 입력 데이터가 독립적이며 모델의 정확도도 높을 가능성
    - s: 특이값 (sigular value)

In [6]:
X = ((np.arange(9)-2)*3).reshape(3,3) #임의의 행렬 X
X

array([[-6, -3,  0],
       [ 3,  6,  9],
       [12, 15, 18]])

In [7]:
y = np.array([[2],[4],[1]]) # 임의의 행렬 y

In [8]:
x, resid, rank, s = np.linalg.lstsq(X,y) #X, y에 대한 최소차증문제 풀기
print(x, '\n\n잔차제곱합:',resid, '\n\n랭크:',rank, '\n\n특이값:',s)

[[-0.46296296]
 [-0.01851852]
 [ 0.42592593]] 

잔차제곱합: [] 

랭크: 2 

특이값: [2.88526380e+01 5.61473790e+00 1.00534292e-16]


C:\Users\chris\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  """Entry point for launching an IPython kernel.


##### 연습문제 2.4.5

In [9]:
from sklearn.datasets import load_boston # 보스턴 집값 문제: 보스턴 내 각 지역(town)의 주택가격에 대한 예측 문제
boston = load_boston()
X = boston.data
y = boston.target
A = X[:4, [0, 4, 5, 6]]  # 'CRIM(범죄율)', 'NOX(공기오염도)', 'RM(방의개수)', 'AGE(오래된 정도)'
b = y[:4]

In [10]:
X.shape # dataset 크기 확인

(506, 13)

In [11]:
A # 임의의 행렬 A 추출

array([[6.320e-03, 5.380e-01, 6.575e+00, 6.520e+01],
       [2.731e-02, 4.690e-01, 6.421e+00, 7.890e+01],
       [2.729e-02, 4.690e-01, 7.185e+00, 6.110e+01],
       [3.237e-02, 4.580e-01, 6.998e+00, 4.580e+01]])

In [12]:
b # 행렬 A에 대한 출력 데이터 벡터 추출

array([24. , 21.6, 34.7, 33.4])

In [13]:
np.linalg.lstsq(A,b) # 임의 추출한 A, b에 대한 모델링

C:\Users\chris\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  """Entry point for launching an IPython kernel.


(array([-3.12710043e+02, -1.15193942e+02,  1.44996465e+01, -1.13259317e-01]),
 array([], dtype=float64),
 4,
 array([1.28391009e+02, 2.97333579e+00, 7.58898410e-02, 7.51932881e-03]))

In [14]:
np.linalg.lstsq(X,y) # 전체 데이터셋인 X,y에 대한 모델링

C:\Users\chris\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  """Entry point for launching an IPython kernel.


(array([-9.28965170e-02,  4.87149552e-02, -4.05997958e-03,  2.85399882e+00,
        -2.86843637e+00,  5.92814778e+00, -7.26933458e-03, -9.68514157e-01,
         1.71151128e-01, -9.39621540e-03, -3.92190926e-01,  1.49056102e-02,
        -4.16304471e-01]),
 array([12228.04626104]),
 13,
 array([1.25851816e+04, 3.44597406e+03, 6.45757109e+02, 4.02050461e+02,
        1.58964612e+02, 1.21502936e+02, 9.04652420e+01, 7.79311708e+01,
        6.50828345e+01, 2.46251803e+01, 1.80945451e+01, 5.51505065e+00,
        1.48096916e+00]))

### 미지수의 수와 방정식의 수

$X \in R^{N*M}$ 일 때,
- 행의 개수($N$): 방정식의 개수, 데이터의 개수
- 열의 개수($M$): 미지수의 개수

1. $N = M$
    - 지금까지 예제로 보았던 경우
    - $X$가 정방행렬이기 때문에 역행렬 계산 및 적용 등에 무리 없음
    
    
2. $N < M$
    - 해가 무수히 많이 존재하기 때문에 방정식을 풀 수 없음.
    - 즉 모델링을 할 수 없음

3. $N > M$
    - 실제로 접하는 대부분의 예측 문제들은 상당히 많은 데이터(=방정식)를 다루기 때문에 대부분의 예측문제는 이 경우에 해당할 것
    - 모든 데이터가 완벽히 선형에 있지 않는 이상, 모든 방정식을 만족하는 해가 존재하지 않게됨

### 최소차승문제 (least squre problem)

$N > M$ 일 때 선립 연립방정식을 구하는 방식으로는 해를 구할 수 없다. (불능)


\begin{align}
\begin{matrix}
x_1 & + & x_2 &   &     & = & 2  \\
    &   & x_2 & + & x_3 & = & 2  \\
x_1 & + & x_2 & + & x_3 & = & 3  \\
x_1 & + & x_2 & + & 2x_3 & = & 4.1  \\
\end{matrix}
\tag{2.4.37}
\end{align}

그럴 때는 차선으로 "좌변(예측값)과 우변(목푯값)을 완벽하게 일치"시키는 것에서 "좌변(예측값)과 우변(목푯값)을 거의 비슷하게 일치"시키는 것으로 문제를 바꿔서 생각할 수 있다.
- "좌변(예측값)과 우변(목푯값)을 완벽하게 일치"시키는 문제 $\rightarrow$ **선형 연립방정식**
- "좌변(예측값)과 우변(목푯값)을 거의 비슷하게 일치"시키는 문제 $=$ "좌변과 우변의 차이를 최소화"시키는 문제 $\rightarrow$ **최소차승문제**


\begin{align}
\begin{matrix}
x_1 & + & x_2 &   &     & = & 2 & &\\
    &   & x_2 & + & x_3 & = & 2 & &\\
x_1 & + & x_2 & + & x_3 & = & 3 & &\\
x_1 & + & x_2 & + & 2x_3 & = & 4 & \approx & 4.1
\end{matrix}
\tag{2.4.38}
\end{align}

- 예측값과 목푯값의 차이 $\rightarrow$ 잔차(residual)

$$e = Ax - b  $$

$$e = Xw - \hat{y} $$

- 잔차(residual)은 벡터이기 때문에, "잔차의 크기를 최소화"한다는 것은 곧 "벡터의 놈(Norm)을 최소화"한다는 것을 의미한다.
    - *trace, determinant는 정방행렬(square matrix)에만 적용 가능하다.*

- 놈(Norm)의 성질에 의해 "norm을 최소화"하는 문제는 "norm의 제곱을 최소화"하는 것과 같다.

위의 내용에 따라 최소차승문제(least square problem)의 정의는 다음과 같다.

\begin{align}
x = \text{arg} \min_x e^Te = \text{arg} \min_x  \; (Ax-b)^T(Ax-b)  
\tag{2.4.41}
\end{align}

\begin{align}
Ax \approx b 
\tag{2.4.42}
\end{align}

\begin{align}
A^TAx \approx A^Tb 
\tag{2.4.43}
\end{align}

- 비록 $A$가 정방행렬이 아니라도, $A^TA$는 정방행렬이다.

- 정방행렬 $A^TA$의 역행렬 $(A^TA)^{-1}$이 존재한다고 가정하면,

\begin{align}
(A^TA)^{-1}(A^TA)x \approx (A^TA)^{-1}A^Tb 
\tag{2.4.44}
\end{align}



\begin{align}
x \approx ((A^TA)^{-1}A^T) b  
\tag{2.4.45}
\end{align}

- 엄밀한 증명은 행렬의 미분과 최적화 개념 필요

- 행렬 A의 의사역행렬(pseudo inverse)
    - 위 식에서 나온 $((A^TA)^{-1}A^T)$를 pseudo inverse라고 정의함
    - 줄여서 $A^+$로 표시할 수 있음

\begin{align}
A^{+} = (A^TA)^{-1}A^T 
\tag{2.4.46}
\end{align}

\begin{align}
x \approx A^+ b  
\tag{2.4.47}
\end{align}

##### 정리하자면,

$X in R^{N*M}$ 일 때,$Xw = \hat{y}\$의 풀이는 다음과 같다.

1. $N = M$, $w = X^{-1}\hat{y}$
2. $N > M$, $w \approx X^{+}\hat{y}$
3. $N < M$, $w$: 부정